### PR TITLE
feat: add license resource

### LINKED ISSUE #5: Add license management capability to coderd provider
Should allow adding / removing licenses, e.g.:

```
resource "coderd_license" "license" {
  license = kubernetes_secret.license.data["license.json"]
}
```

### PR DESCRIPTION
**Pull request by ethanndickson**
Closes #5.

Even if we're not totally convinced of the value of this, and whether it'll get used, I figured this would be worth doing before the next release - the implementation was also pretty straight-forward, and didn't take long.

### CODE REVIEW

**Comment by ethanndickson** [comment]
* **#99** <a href="https://app.graphite.dev/github/pr/coder/terraform-provider-coderd/99?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ethanndickson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/provider/license_resource.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/coder/coder/v2/codersdk"
```
Comment (deansheather) [inline internal/provider/license_resource.go:8]: coder imports should go in their own section at the bottom

File: internal/provider/license_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+func (c testAccLicenseResourceconfig) String(t *testing.T) string {
+	t.Helper()
+	tpl := `
+provider coderd {
+	url   = "{{.URL}}"
+	token = "{{.Token}}"
+}
+
+resource "coderd_license" "test" {
+	license              = "{{.License}}"
```
Comment (deansheather) [inline internal/provider/license_resource_test.go:60]:
```suggestion
	license = "{{.License}}"
```

File: internal/provider/license_resource.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+			found = true
+			expiresAt, err := license.ExpiresAt()
+			if err != nil {
+				resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to parse license expiration, got error: %s", err))
+				return
+			}
+			if expiresAt.Before(time.Now()) {
+				resp.Diagnostics.AddError("Client Error", "License has expired.")
+				return
+			}
```
Comment (deansheather) [inline internal/provider/license_resource.go:149]: I think this check should be removed, we don't check expiry on the server except for computing entitlements. You can insert and view expired licenses AFAIK

File: internal/provider/license_resource.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	License   types.String `tfsdk:"license"`
+}
+
+func (r *LicenseResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_license"
+}
+
+func (r *LicenseResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "A license for a Coder deployment.\n\nTerraform does not guarantee this resource" +
```
Comment (deansheather) [inline internal/provider/license_resource.go:42]: You should recommend using multiple license resources if you're trying to swap out a license, as terraform will delete the old one THEN create the new one when replacing which will result in a brief unlicensed period

**Review by deansheather — APPROVED** [review]

**Pull request merged**